sycl/source/device_selector.cpp

```diff
@@ -10,81 +10,129 @@
 #include <CL/sycl/device_selector.hpp>
 #include <CL/sycl/exception.hpp>
 #include <CL/sycl/stl.hpp>
+#include <detail/device_impl.hpp>
 #include <detail/force_device.hpp>
 // 4.6.1 Device selection class
 
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
+
+// Utility function to check if device is of the preferred SYCL_BE.
+static bool isDeviceOfPreferredSyclBe(const device &Device) {
+  if (Device.is_host())
+    return false;
+
+  detail::pi::Backend PreferredBE = detail::pi::getPreferredBE();
+  detail::pi::Backend DeviceBE =
+      detail::getSyclObjImpl(Device)->getPlugin().getBackend();
+  return PreferredBE == DeviceBE ? true : false;
+}
+
 device device_selector::select_device() const {
   vector_class<device> devices = device::get_devices();
   int score = -1;
   const device *res = nullptr;
-  for (const auto &dev : devices)
-    if (score < operator()(dev)) {
+  for (const auto &dev : devices) {
+    int dev_score = operator()(dev);
+    if (detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_ALL)) {
+      string_class PlatformVersion = dev.get_info<info::device::platform>()
+                                         .get_info<info::platform::version>();
+      string_class DeviceName = dev.get_info<info::device::name>();
+      std::cout << "SYCL_PI_TRACE[1]: select_device(): -> score = " << score
+                << std::endl
+                << "SYCL_PI_TRACE[1]:   platform: " << PlatformVersion
+                << std::endl
+                << "SYCL_PI_TRACE[1]:   device: " << DeviceName << std::endl;
+    }
+
+    if (score < dev_score) {
       res = &dev;
-      score = operator()(dev);
+      score = dev_score;
```

Comment:
This reminds me that we have never said anything about the fact that the device selector should always return the same value for the same device... ;-) At least calling it once and caching the result like here is safer.

Reply:
Well, at some point that *was* intentional :-)

Reply:
It is legal probably and might still be intentional. For example imagine you want to implement the `random_selector`. :-)

At least we want the selector to be called only once per device by the SYCL runtime inside 1 SYCL consumption or some havoc can happen...